File: 001-scaler-academy/001-accelerate-your-tech-career/003-data-structure/001-array/001-Rain Water Trapped/Solution.py
class Solution:
	# @param A : tuple of integers
	# @return an integer
	def trap(self, A):
		n = len(A)
		max_left_height=[None]*n
		max_right_height=[None]*n

		# Scanning through the list to find the max_left_height for each gap
		current_left_height = 0
		for i in range(n):
			max_left_height[i] = current_left_height
			current_left_height = max(current_left_height, A[i])

		current_right_height = 0
		for i in range(n):
			max_right_height[n-i] = current_right_height
			current_right_height = Math.max(current_right_height, A[n-i])

		total_water_trapped = 0
		for i in range(n):
			total_water_trapped += max(0, min(max_left_height[i], max_right_height[i]) - A[i])

		return total_water_trapped

File: 001-scaler-academy/001-accelerate-your-tech-career/003-data-structure/001-array/001-Rain Water Trapped/Solution.c
/**
 * @input A : Read only ( DON'T MODIFY ) Integer array
 * @input n : Integer array's ( A ) length
 *
 * @Output Integer
 */
int trap(const int* A, int n) {
        int maxLeftHeight[n];
        int maxRightHeight[n];

        // Scanning through the array to find the maxLeftHeight for each gap
        int currentLeftHeight = 0,i;
        for (i = 0; i < n; i++) {
            maxLeftHeight[i] = currentLeftHeight;
            currentLeftHeight = max(currentLeftHeight,A[i]);
        }

        // Scanning through the array to find the maxRightHeight for each gap
        int currentRightHeight = 0;
        for (i = n - 1; i >= 0; i--) {
            maxRightHeight[i] = currentRightHeight;
            currentRightHeight = max(currentRightHeight,A[i]);
        }

        // Calculating the total water trapped within the gaps
        int totalWaterTrapped=0;
        for (i = 0; i < n; i++) {
            totalWaterTrapped += max(0, min(maxLeftHeight[i], maxRightHeight[i]) - A[i]);
        }
        return totalWaterTrapped;
    }

int min(int a, int b){
    return a<=b?a:b;
}

int max(int a, int b){
    return a>=b?a:b;
}